### PR TITLE
Add Update to Environments

### PR DESCRIPTION
**Pull request by tlake**
**What does this pull request do?**
Wires up EnvironmentUpdate functionality to the API.

**Checklist**
- [x] Unit tests
- ~Smoke tests (if applicable)~
- ~System tests (if applicable)~
- ~Documentation (if applicable)~

### CODE REVIEW

**Comment by tlake** [comment]
I went back and forth for a while with patterns for the PUT request, which took up a lot of time, so I've just decided to stop playing and throw it out for other people to tear apart. Currently, I envision the PUT request hitting the `environment/{id}/` endpoint, so the target environment id is a path variable. Then the body of the PUT request should contain the items to update, which I think should just be the min cluster count. I changed `EnvironmentProvider.Update()` to take a `models.UpdateEnvironmentRequest` object instead of `environmentID string, minSize int64`, and before the request is sent to the scheduler, it's populated with the id from the path.

**Comment by zpatrick** [comment]
Because we need the `EntityID` in the object model for it to be properly picked up and executed by a job runner, I think we should put this request into `PUT /entity` instead of `PUT /entity/{id}`. This avoids any confusion between the request field and the path param.

File: api/job/types.go [hunk 1]
```diff
@@ -12,6 +12,7 @@ const (
 	CreateLoadBalancerJob JobType = "CreateLoadBalancer"
 	CreateServiceJob      JobType = "CreateService"
 	CreateTaskJob         JobType = "CreateTask"
+	UpdateEnvironmentJob  JobType = "UpdateEnvironment"
```
[the remark this answers is not in the document]
Comment (tlake) [reply api/job/types.go]: ✅

File: api/provider/environment.go [hunk 2]
```diff
@@ -6,5 +6,6 @@ type EnvironmentProvider interface {
 	Create(req models.CreateEnvironmentRequest) (*models.Environment, error)
 	Read(environmentID string) (*models.Environment, error)
 	List() ([]models.EnvironmentSummary, error)
+	Update(req models.UpdateEnvironmentRequest) (*models.Environment, error)
```
[the remark this answers is not in the document]
Comment (tlake) [reply api/provider/environment.go]: ✅

File: common/models/update_environment_request.go [hunk 3]
```diff
@@ -1,5 +1,29 @@
...
+	swagger "github.com/zpatrick/go-plugin-swagger"
+)
+
 type UpdateEnvironmentRequest struct {
-	MinClusterCount int `json:"min_cluster_count"`
+	EnvironmentID   string `json:"environment_id"`
+	MinClusterCount int64  `json:"min_cluster_count"`
+}
+
+func (r UpdateEnvironmentRequest) Validate() error {
```
[the remark this answers is not in the document]
Comment (tlake) [reply common/models/update_environment_request.go]: ✅

File: common/models/update_environment_request.go [hunk 4]
```diff
@@ -1,5 +1,29 @@
...
+import (
+	"fmt"
+
+	swagger "github.com/zpatrick/go-plugin-swagger"
+)
+
 type UpdateEnvironmentRequest struct {
-	MinClusterCount int `json:"min_cluster_count"`
+	EnvironmentID   string `json:"environment_id"`
+	MinClusterCount int64  `json:"min_cluster_count"`
```
[the remark this answers is not in the document]
Comment (tlake) [reply common/models/update_environment_request.go]: Changing this to `*int` for ease of use; int64 nonsense is handled in the `provider` package.

✅

File: api/controllers/environment.go [hunk 5]
```diff
@@ -78,3 +79,19 @@ func (e *EnvironmentController) ListEnvironments(c *fireball.Context) (fireball.
 	return fireball.NewJSONResponse(200, summaries)
 
 }
+
+func (e *EnvironmentController) UpdateEnvironment(c *fireball.Context) (fireball.Response, error) {
+	id := c.PathVariables["id"]
```
[the remark this answers is not in the document]
Comment (tlake) [reply api/controllers/environment.go]: ✅

File: api/controllers/environment_test.go [hunk 6]
```diff
@@ -152,3 +152,42 @@ func TestListEnvironments(t *testing.T) {
...
+
+func TestUpdateEnvironment(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	mockEnvironmentProvider := mock_provider.NewMockEnvironmentProvider(ctrl)
+	mockJobScheduler := mock_job.NewMockScheduler(ctrl)
+	controller := NewEnvironmentController(mockEnvironmentProvider, mockJobScheduler)
+
+	httpReq := models.UpdateEnvironmentRequest{
```
[the remark this answers is not in the document]
Comment (tlake) [reply api/controllers/environment_test.go]: ✅

File: common/models/update_environment_request.go [hunk 7]
```diff
@@ -1,5 +1,29 @@
...
+}
+
+func (r UpdateEnvironmentRequest) Validate() error {
+	if r.MinClusterCount < 0 {
```
[the remark this answers is not in the document]
Comment (tlake) [reply common/models/update_environment_request.go]: Hm, something went wonky somewhere - that was in `Validate` already at some point. It should be back there now.

✅

File: common/models/update_environment_request.go [hunk 8]
```diff
@@ -1,5 +1,29 @@
...
+	}
+
+	return nil
+}
+
+func (e UpdateEnvironmentRequest) Definition() swagger.Definition {
+	return swagger.Definition{
+		Type: "object",
+		Properties: map[string]swagger.Property{
+			"min_cluster_count": swagger.NewIntProperty(),
```
[the remark this answers is not in the document]
Comment (tlake) [reply common/models/update_environment_request.go]: Again, this was also in there at one point; in the most recent version now.

✅

**Pull request merged**